### PR TITLE
test: more automatic tests for migrations

### PR DESCRIPTION
**Pull request by crepererum**
This is the final piece for #7897. Migrations should now be hardened enough.

Note that some edge cases are NOT really testable (to my knowledge), esp. `CREATE INDEX CONCURRENTLY`. I'm not aware of any way to get an index into an invalid state w/o messing with the postgres progress directly. However #8394 would detect this at runtime and we are able to fix migrations as shown in #8450. So I think this is "good enough".

### CODE REVIEW

File: iox_catalog/src/migrate.rs [hunk 1]
```diff
@@ -872,6 +872,105 @@ ORDER BY pg_class.relname
...
+
+    /// Test migration.
+    ///
+    /// This runs the migrations to check if they pass. The given factory must provide an empty schema (i.e. w/o any
+    /// migrations applied).
+    ///
+    /// # Tests
+    /// This tests that:
+    ///
+    /// - **run ones:** All migrations work when ran ones.
```
Comment (mhilton) [inline iox_catalog/src/migrate.rs:890]: Do you mean "once" rather than "ones" here?

**Review by mhilton — APPROVED** [review]

File: iox_catalog/src/migrate.rs (hunk 1 above)
Comment (crepererum) [reply iox_catalog/src/migrate.rs:890]: yes, I always confuse these :sweat_smile:
